delta-lake/delta-20x/src/main/scala/org/apache/spark/sql/delta/rapids/delta20x/GpuMergeIntoCommand.scala

```diff
@@ -0,0 +1,1167 @@
+        createMetric(sc, "number of source rows (during repeated scan)"),
+    "numTargetRowsCopied" -> createMetric(sc, "number of target rows rewritten unmodified"),
+    "numTargetRowsInserted" -> createMetric(sc, "number of inserted rows"),
+    "numTargetRowsUpdated" -> createMetric(sc, "number of updated rows"),
+    "numTargetRowsDeleted" -> createMetric(sc, "number of deleted rows"),
+    "numTargetFilesBeforeSkipping" -> createMetric(sc, "number of target files before skipping"),
+    "numTargetFilesAfterSkipping" -> createMetric(sc, "number of target files after skipping"),
+    "numTargetFilesRemoved" -> createMetric(sc, "number of files removed to target"),
+    "numTargetFilesAdded" -> createMetric(sc, "number of files added to target"),
+    "numTargetChangeFilesAdded" ->
+        createMetric(sc, "number of change data capture files generated"),
+    "numTargetChangeFileBytes" ->
+        createMetric(sc, "total size of change data capture files generated"),
+    "numTargetBytesBeforeSkipping" -> createMetric(sc, "number of target bytes before skipping"),
+    "numTargetBytesAfterSkipping" -> createMetric(sc, "number of target bytes after skipping"),
+    "numTargetBytesRemoved" -> createMetric(sc, "number of target bytes removed"),
+    "numTargetBytesAdded" -> createMetric(sc, "number of target bytes added"),
+    "numTargetPartitionsAfterSkipping" ->
+        createMetric(sc, "number of target partitions after skipping"),
+    "numTargetPartitionsRemovedFrom" ->
+        createMetric(sc, "number of target partitions from which files were removed"),
+    "numTargetPartitionsAddedTo" ->
+        createMetric(sc, "number of target partitions to which files were added"),
+    "executionTimeMs" ->
+        createMetric(sc, "time taken to execute the entire operation"),
+    "scanTimeMs" ->
+        createMetric(sc, "time taken to scan the files for matches"),
+    "rewriteTimeMs" ->
+        createMetric(sc, "time taken to rewrite the matched files"))
+
+  override def run(spark: SparkSession): Seq[Row] = {
+    if (migratedSchema.isDefined) {
+      // Block writes of void columns in the Delta log. Currently void columns are not properly
+      // supported and are dropped on read, but this is not enough for merge command that is also
+      // reading the schema from the Delta log. Until proper support we prefer to fail merge
+      // queries that add void columns.
+      val newNullColumn = SchemaUtils.findNullTypeColumn(migratedSchema.get)
+      if (newNullColumn.isDefined) {
+        throw new AnalysisException(
+          s"""Cannot add column '${newNullColumn.get}' with type 'void'. Please explicitly specify a
+             |non-void type.""".stripMargin.replaceAll("\n", " ")
+        )
+      }
+    }
+
+    recordDeltaOperation(targetDeltaLog, "delta.dml.merge") {
+      val startTime = System.nanoTime()
+      gpuDeltaLog.withNewTransaction { deltaTxn =>
+        if (target.schema.size != deltaTxn.metadata.schema.size) {
```

Comment:
Yeah this is a bit nasty, a whole new feature added to a patch release of Delta Lake.  I'll update the docs to say we only support 2.0.1, 2.1.1 and 2.2.0 rather than the ".x" versions, and I filed #7582 to track.